helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -200,6 +206,45 @@ protected void handleEvent(NotificationContext event) {
         logger.warn("Exception in callback processing thread. Skipping callback", e);
       }
     }
+
+    public boolean queueIsEmpty() {
+      return _eventQueue.isEmpty();
+    }
+  }
+
+  /**
+   * This class represents each periodic refresh task
+   * If the deadline passed, enqueue an event to do a refresh
+   * If the deadline has not passed (some events happened in between so don't need to do a refresh),
+   * wait until reaching the interval and check again
+   */
+  class TriggerTask implements Runnable {
+    @Override
+    public void run() {
+      try {
+        while (true) {
+          long currentTime = System.currentTimeMillis();
+          long remainingTime = _lastInvokeTime + _periodicTriggerInterval - currentTime;
+          if (remainingTime <= 0) {
+            // If there is no event in the queue, meaning this long idle time could be due to lack of events
+            // Otherwise, if there is event in the queue, this long idle time is due to slow process of events
+            if (_batchCallbackProcessor == null || _batchCallbackProcessor.queueIsEmpty()) {
```

Comment:
If _batchCallbackProcessor is not empty, but the callback handling takes a while. then this loop will do a busy loop, right?
More graceful design is to wait until the queue becomes non-empty then loop to check the remaining time again.

Reply:
I don't think it will deadlock. But this also works.